File: test_drag.py
from playwright.sync_api import sync_playwright
import time

def run(playwright):
    browser = playwright.chromium.launch(headless=True)
    context = browser.new_context()
    page = context.new_page()

    print("Navigating to mission.html...")
    page.on("console", lambda msg: print(f"Browser Console: {msg.text}"))
    page.goto("http://localhost:8080/mission.html")
    page.wait_for_timeout(2000)
    
    print("Taking initial screenshot...")
    page.screenshot(path="test_drag_init.png")

    html = page.content()
    with open("test_drag_dom.html", "w") as f:
        f.write(html)
        
    rows = page.locator(".blocklyTreeRow").all()
    print(f"Found {len(rows)} tree rows in total.")
    for i, r in enumerate(rows):
        print(f"Row {i} classes: {r.get_attribute('class')} visible: {r.is_visible()}")

    print("Clicking first category via JS dispatch...")
    row_handle = page.locator(".blocklyTreeRow").nth(1)
    row_handle.evaluate("el => el.dispatchEvent(new PointerEvent('pointerdown', {bubbles: true}))")
    page.wait_for_timeout(100)
    row_handle.evaluate("el => el.dispatchEvent(new PointerEvent('pointerup', {bubbles: true}))")
    row_handle.evaluate("el => el.click()")
    
    page.wait_for_timeout(1500)
    
    print("Taking category open screenshot...")
    page.screenshot(path="test_drag_cat_open.png")

    # Check if blocks appeared in flyout
    blocks = page.locator(".blocklyFlyout .blocklyDraggable")
    count = blocks.count()
    print(f"Found {count} draggable blocks in flyout.")

    if count > 0:
        block = blocks.first
        box = block.bounding_box()
        print(f"Block bounding box: {box}")
        
        if box:
            print("Dragging block to workspace...")
            # Move to center of block
            page.mouse.move(box["x"] + box["width"] / 2, box["y"] + box["height"] / 2)
            page.mouse.down()
            
            # Move to workspace area (e.g. 500, 300)
            page.mouse.move(500, 300, steps=10)
            page.mouse.up()
            
            page.wait_for_timeout(1000)
            page.screenshot(path="test_drag_after_drop.png")
            
            # Evaluate blocks in workspace
            ws_blocks = page.evaluate("Blockly.getMainWorkspace().getAllBlocks().length")
            print(f"Blocks in workspace after drop: {ws_blocks}")

    context.close()
    browser.close()

with sync_playwright() as playwright:
    run(playwright)

File: test_drag.js
const puppeteer = require('puppeteer');

(async () => {
    const browser = await puppeteer.launch({ headless: "new" });
    const page = await browser.newPage();
    try {
        await page.goto('http://localhost:8080/mission.html', { waitUntil: 'networkidle0' });
        
        // Take an initial screenshot
        await page.screenshot({ path: 'test_drag_init.png' });
        
        // Wait for Blockly to load
        await page.waitForFunction(() => typeof Blockly !== 'undefined' && Blockly.getMainWorkspace() !== null);
        
        // Find the "ACTION" category row and click it
        const actionCat = await page.$('.blocklyTreeRow');
        if (actionCat) {
            console.log("Found category, clicking...");
            await actionCat.click();
            await page.waitForTimeout(500);
            
            // Take a screenshot after opening category
            await page.screenshot({ path: 'test_drag_cat_open.png' });
            
            // Try to find a block in the flyout
            const blocks = await page.$$('.blocklyFlyout .blocklyDraggable');
            console.log("Found draggable blocks in flyout:", blocks.length);
            
            if (blocks.length > 0) {
                const blockBox = await blocks[0].boundingBox();
                console.log("Block bounding box:", blockBox);
                
                // Try to drag the block to the workspace (e.g. at x: 400, y: 200)
                await page.mouse.move(blockBox.x + blockBox.width / 2, blockBox.y + blockBox.height / 2);
                await page.mouse.down();
                await page.mouse.move(400, 200, { steps: 10 });
                await page.mouse.up();
                
                await page.waitForTimeout(500);
                await page.screenshot({ path: 'test_drag_after_drop.png' });
                
                // Check if block was added to workspace
                const workspaceBlocks = await page.evaluate(() => Blockly.getMainWorkspace().getAllBlocks().length);
                console.log("Blocks in workspace after drop:", workspaceBlocks);
            }
        } else {
            console.log("No category found.");
        }
        
    } catch (e) {
        console.error("Error:", e);
    } finally {
        await browser.close();
    }
})();
